client/src/lib.rs

```diff
@@ -0,0 +1,55 @@
+use fuel_debugger::{Command, Response};
+use std::io;
+use std::net::SocketAddr;
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::net::{
+    tcp::{OwnedReadHalf, OwnedWriteHalf},
+    TcpListener,
+};
+
+pub mod names;
+
+pub struct Listener {
+    listener: TcpListener,
+}
+
+impl Listener {
+    pub fn new(listener: TcpListener) -> Self {
+        Self { listener }
+    }
+
+    pub async fn accept(&self) -> io::Result<(Client, SocketAddr)> {
+        let (socket, addr) = self.listener.accept().await?;
+
+        let (reader, writer) = socket.into_split();
+        let reader = BufReader::new(reader);
+
+        Ok((Client { reader, writer }, addr))
+    }
+}
+
+pub struct Client {
+    reader: BufReader<OwnedReadHalf>,
+    writer: OwnedWriteHalf,
+}
+
+impl Client {
+    pub async fn cmd(&mut self, cmd: &Command) -> io::Result<Response> {
+        let mut v = serde_json::to_string(cmd).expect("Serialization failed");
+        v.push('\n');
+        self.writer
+            .write_all(v.as_bytes())
+            .await
+            .expect("Sending failed");
+
+        let mut line = String::new();
+        let _ = self.reader.read_line(&mut line).await?;
+        if line.is_empty() {
+            Err(io::Error::new(io::ErrorKind::Other, "Connection closed"))
+        } else {
+            let resp: Response =
```

Comment:
This response isn't always guaranteed to be related to the preceding command. For instance, if single-step debugging is disabled, you could receive a `Breakpoint` response at any time which might not be related to the current command (like `ReadRegisters`). 

Requests should have a `request_id` generated by the client, and the server responses should set an optional `request_id` field if applicable. I think we should have a background task which constantly polls responses and stores them into a map (`responses: RwLock<HashMap<RequestId, Response>>`) for later retrieval.

Reply:
I don't think this is correct. The commands must be processed sequentially to make debugging deterministic. If it was more complicated, using an RPC library would make sense.

Reply:
Without single step debugging enable, couldn't this client receive a message any time a breakpoint was hit, regardless of any outstanding requests? Seems like this could happen if the breakpoint response was sent just before or around the same time as the command was received by the server?

Reply:
The VM is only ever ran in blocking mode, with sent commands processed only before exection and when stopped at a breakpoint. Each command is processed and response sent before the next one is read from the input buffer. The other side of the socket, which drives the VM, looks essentially like this:

```rust
loop {
  command = socket.recvline();
  if command == Command::Continue {
    break;
  } else {
    response = run_command(command, &mut vm);
    socket.sendline(response);
  }
}
loop {
  vm_state = vm.run_until_next_breakpoint_blocking();
  match vm_state {
    Breakpoint => {
      socket.send(vm_state.breakpoint_info());
      loop {
        command = socket.recvline()
        if command == Command::Continue {
          break;
        } else {
          response = run_command(command, &mut vm);
          socket.sendline(response);
        }
      }
    },
    Terminated => break,
  }
}
socket.send(vm_state);
```

Reply:
The DAP protocol for instance uses JSONRPC. For every request, we have a well-formed tracking id.

Assuming the request/response is synchronous for async functions is a big decision - and unnecessary if we just track the requests with IDs.

Reply:
I can see the design flaw, now.